README
App-Environ version 0.18
========================

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DESCRIPTION

App::Environ is the simple environment to build applications using service
locator pattern. Allows register different application components that provide
common resources.

AUTHOR

Eugene Ponizovsky, E<lt>ponizovsky@gmail.comE<gt>

COPYRIGHT AND LICENSE

Copyright (c) 2016, Eugene Ponizovsky, <ponizovsky@gmail.com>.
All rights reserved.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.


Changes
Revision history for Perl extension App::Environ.

0.20 Thu Feb 16 14:51:05 MSK 2017
  - Added handling of environment variable APPCONF_EXPORT_ENV.
  - Added description for all environment variables used by
    App::Environ::Config.
  - Added README.pod file instead of README file.

0.18 Thu Dec 8 18:06:30 MSK 2016
  - FEATURE: Changed concept of asynchronous processing.
    Backward compatible was broken.

0.16 Thu Dec 1 17:34:05 MSK 2016
  - BUGFIX: Fixed issue with repeatable initialization/finalization in
    App::Environ::Config.

0.14 Wed Nov 30 17:07:55 MSK 2016
  - BUGFIX: Fixed error handling for asynchronous mode.

0.12 Fri Nov 25 12:09:30 MSK 2016
  - Removed garbage from example.

0.10 Fri Nov 25 12:03:05 MSK 2016
  - Fixes in POD.
  - Light refactoring.

0.08 Wed Nov 23 17:17:05 MSK 2016
  - FEATURE: Added ability to handle errors in asynchronous mode.
  - BUGFIX: Fixed handling of unknown events in asynchronous mode.
  - Improved POD.
  - Improved test.

0.06 Tue Nov 22 12:22:40 MSK 2016
  - Removed "finalize:r" handler from App::Environ.
  - Refactoring.

0.04 Mon Nov 21 14:30:20 MSK 2016
  - Added "finalize:r" handler in App::Environ.
  - Do not die if handler has already registered.
  - Fixed handling of ENV variables in App::Environ::Config.

0.02 Wed Nov 16 16:11:20 MSK 2016
	- Initial release.


t/lib/Bar.pm
package Bar;

use strict;
use warnings;

use App::Environ;
use App::Environ::Config;
use Carp qw( croak );

App::Environ::Config->register( qw( bar.yml ) );

App::Environ->register( __PACKAGE__,
  initialize     => sub { __PACKAGE__->_initialize(@_) },
  reload         => sub { __PACKAGE__->_reload(@_) },
  'wait_async:r' => sub { __PACKAGE__->_wait_async(@_) },
  'finalize:r'   => sub { __PACKAGE__->_finalize(@_) },
);

my $INSTANCE;


sub instance {
  unless ( defined $INSTANCE ) {
    croak __PACKAGE__ . ' must be initialized first';
  }

  return $INSTANCE;
}

sub _initialize {
  my $class = shift;
  my @args  = @_;

  my $bar_config = App::Environ::Config->instance->{'bar'};

  $INSTANCE = {
    config    => $bar_config,
    init_args => \@args,
    reloads   => 0,
  };

  return;
}

sub _reload {
  $INSTANCE->{config} = App::Environ::Config->instance->{'bar'};
  $INSTANCE->{reloads}++;

  return;
}

sub _wait_async {
  my $class    = shift;
  my $need_err = shift;
  my $cb       = shift;

  $cb->();

  return;
}

sub _finalize {
  undef $INSTANCE;
  return;
}

1;


t/lib/Foo.pm
package Foo;

use strict;
use warnings;

use App::Environ;
use App::Environ::Config;
use Carp qw( croak );

App::Environ::Config->register( qw( foo.json ) );

App::Environ->register( __PACKAGE__,
  initialize     => sub { __PACKAGE__->_initialize(@_) },
  reload         => sub { __PACKAGE__->_reload(@_) },
  'wait_async:r' => sub { __PACKAGE__->_wait_async(@_) },
  'finalize:r'   => sub { __PACKAGE__->_finalize(@_) },
);

my $INSTANCE;


sub instance {
  unless ( defined $INSTANCE ) {
    croak __PACKAGE__ . ' must be initialized first';
  }

  return $INSTANCE;
}

sub _initialize {
  my $class = shift;
  my @args  = @_;

  my $foo_config = App::Environ::Config->instance->{'foo'};

  $INSTANCE = {
    config    => $foo_config,
    init_args => \@args,
    reloads   => 0,
  };

  return;
}

sub _reload {
  my $class     = shift;
  my $raise_err = shift;

  die "Some error.\n" if ($raise_err);

  $INSTANCE->{config} = App::Environ::Config->instance->{'foo'};
  $INSTANCE->{reloads}++;

  return;
}

sub _wait_async {
  my $class     = shift;
  my $raise_err = shift;
  my $cb        = shift;

  if ($raise_err) {
    $cb->('Some error.');
    return;
  }

  $cb->();

  return;
}

sub _finalize {
  undef $INSTANCE;
  return;
}

1;


examples/async.pl
#!/usr/bin/env perl

use 5.008000;
use strict;
use warnings;

use lib './examples/lib';

use FindBin;
use App::Environ;
use Data::Dumper;

BEGIN {
  $ENV{APPCONF_DIRS} = "$FindBin::Bin/etc";
}

use Cat;
use Dog;
use Cow;

use AnyEvent;

my $cv = AE::cv;

App::Environ->send_event( 'initialize', qw( foo bar ),
  sub {
    my $err = shift;

    if ( defined $err ) {
      $cv->croak($err);
      return;
    }

    $cv->send
  }
);

$cv->recv;

my $cat_inst = Cat->instance;
my $dog_inst = Dog->instance;
my $cow_inst = Cow->instance;

print Dumper($cat_inst);
print Dumper($dog_inst);
print Dumper($cow_inst);

$cv = AE::cv;

App::Environ->send_event( 'reload',
  sub {
    my $err = shift;

    if ( defined $err ) {
      $cv->croak($err);
      return;
    }

    $cv->send
  }
);

$cv->recv;

$cv = AE::cv;

App::Environ->send_event( 'finalize:r',
  sub {
    my $err = shift;

    if ( defined $err ) {
      $cv->croak($err);
      return;
    }

    $cv->send
  }
);

$cv->recv;


lib/App/Environ.pm
package App::Environ;

use 5.008000;
use strict;
use warnings;

our $VERSION = '0.18';

use Carp qw( croak );

my %REGD_HANDLERS;
my %MODULES_IDX;


sub register {
  my $class        = shift;
  my $module_class = shift;
  my %handlers     = @_;

  unless ( defined $module_class ) {
    croak 'Module class must be specified';
  }

  unless ( exists $MODULES_IDX{$module_class} ) {
    $MODULES_IDX{$module_class} = {};
  }
  my $events_idx = $MODULES_IDX{$module_class};

  while ( my ( $event_name, $handler ) = each %handlers ) {
    next if exists $events_idx->{$event_name};

    $events_idx->{$event_name} = 1;
    unless ( exists $REGD_HANDLERS{$event_name} ) {
      $REGD_HANDLERS{$event_name} = [];
    }
    if ( $event_name =~ m/\:r$/ ) {
      unshift( @{ $REGD_HANDLERS{$event_name} }, $handler );
    }
    else {
      push( @{ $REGD_HANDLERS{$event_name} }, $handler );
    }
  }

  return;
}

sub send_event {
  my $class      = shift;
  my $event_name = shift;
  my $cb         = pop if ref( $_[-1] ) eq 'CODE';
  my @args       = @_;

  unless ( defined $event_name ) {
    croak 'Event name must be specified';
  }

  unless ( exists $REGD_HANDLERS{$event_name} ) {
    $cb->() if defined $cb;
    return;
  }

  my @handlers = @{ $REGD_HANDLERS{$event_name} };
  if ( defined $cb ) {
    $class->_process_async( \@handlers, \@args, $cb );
  }
  else {
    foreach my $handler (@handlers) {
      $handler->(@args);
    }
  }

  return;
}

sub _process_async {
  my $class    = shift;
  my $handlers = shift;
  my $args     = shift;
  my $cb       = shift;

  my $handler = shift @{$handlers};

  $handler->( @{$args},
    sub {
      my $err = shift;

      if ( defined $err ) {
        $cb->($err);
        return;
      }

      if ( @{$handlers} ) {
        $class->_process_async( $handlers, $args, $cb );
        return;
      }

      $cb->();
    }
  );

  return;
}

1;
__END__

=head1 NAME

App::Environ - Simple environment to build applications using service locator
pattern

=head1 SYNOPSIS

  use App::Environ;

  App::Environ->register( __PACKAGE__,
    initialize   => sub { ... },
    reload       => sub { ... },
    'finalize:r' => sub { ... },
  );

  App::Environ->send_event( 'initialize', qw( foo bar ) );
  App::Environ->send_event('reload');
  App::Environ->send_event( 'pre_finalize:r', sub {...} );
  App::Environ->send_event('finalize:r');

=head1 DESCRIPTION

App::Environ is the simple environment to build applications using service
locator pattern. Allows register different application components that provide
common resources.

=head1 METHODS

=head2 register( $class, \%handlers )

The method registers handlers for specified events. When some event have been
sent, registered event handlers will be processed in order in which they was
registered. If you want that event handlers have been processed in reverse
order, add postfix C<:r> to event name. All arguments that have been specified
in C<send_event> method (see below) are passed to called event handler. If in
the last argument is passed the callback, the handler must call it when
processing will be done. If the handler was called with callback and some error
occurred, the callback must be called with error message in first argument.

  App::Environ->register( __PACKAGE__,
    initialize => sub {
      my @args = @_;

      # handling...
    },
  );

=head2 send_event( $event [, @args ] [, $cb->( [ $err ] ) ] )

Sends specified event to App::Environ. All handlers registered for this event
will be processed. Arguments specified in C<send_event> method will be passed
to event handlers. If the callback is passed in the last argument, event
handlers will be processed in asynchronous mode.

  App::Environ->send_event( 'initialize', qw( foo bar ) );

  App::Environ->send_event( 'pre_finalize:r'
    sub {
      my $err = shift;

      if ( defined $err ) {
        # error handling...

        return;
      }

      # success handling...
    }
  );

=head1 SEE ALSO

L<App::Environ::Config>

Also see examples from the package to better understand the concept.

=head1 AUTHOR

Eugene Ponizovsky, E<lt>ponizovsky@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2016, Eugene Ponizovsky, E<lt>ponizovsky@gmail.comE<gt>.
All rights reserved.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut


examples/lib/Cat.pm
package Cat;

use strict;
use warnings;

use App::Environ;
use App::Environ::Config;
use AnyEvent;
use Carp qw( croak );

App::Environ::Config->register( qw( cat.json ) );

App::Environ->register( __PACKAGE__,
  initialize   => sub { __PACKAGE__->_initialize(@_) },
  reload       => sub { __PACKAGE__->_reload(@_) },
  'finalize:r' => sub { __PACKAGE__->_finalize(@_) },
);

my $INSTANCE;


sub instance {
  unless ( defined $INSTANCE ) {
    croak __PACKAGE__ . ' must be initialized first';
  }

  return $INSTANCE;
}

sub _initialize {
  my $class = shift;
  my $cb    = pop if ref( $_[-1] ) eq 'CODE';

  my $cat_config = App::Environ::Config->instance->{'cat'};

  $INSTANCE = {
    config    => $cat_config,
    init_args => [@_],
  };

  print __PACKAGE__ . " initialized\n";

  if ( defined $cb ) {
    AE::postpone { $cb->() };
  }

  return;
}

sub _reload {
  my $cb = pop if ref( $_[-1] ) eq 'CODE';

  $INSTANCE->{config} = App::Environ::Config->instance->{'cat'};

  print __PACKAGE__ . " reloaded\n";

  if ( defined $cb ) {
    AE::postpone { $cb->() };
  }

  return;
}

sub _finalize {
  my $cb = pop if ref( $_[-1] ) eq 'CODE';

  undef $INSTANCE;

  print __PACKAGE__ . " finalized\n";

  if ( defined $cb ) {
    AE::postpone { $cb->() };
  }

  return;
}

1;


examples/lib/Cow.pm
package Cow;

use strict;
use warnings;

use App::Environ;
use App::Environ::Config;
use AnyEvent;
use Carp qw( croak );

App::Environ::Config->register( qw( cow.yml ) );

App::Environ->register( __PACKAGE__,
  initialize   => sub { __PACKAGE__->_initialize(@_) },
  reload       => sub { __PACKAGE__->_reload(@_) },
  'finalize:r' => sub { __PACKAGE__->_finalize(@_) },
);

my $INSTANCE;


sub instance {
  unless ( defined $INSTANCE ) {
    croak __PACKAGE__ . ' must be initialized first';
  }

  return $INSTANCE;
}

sub _initialize {
  my $class = shift;
  my $cb    = pop if ref( $_[-1] ) eq 'CODE';

  my $cow_config = App::Environ::Config->instance->{'cow'};

  $INSTANCE = {
    config    => $cow_config,
    init_args => [@_],
  };

  print __PACKAGE__ . " initialized\n";

  if ( defined $cb ) {
    AE::postpone { $cb->() };
  }

  return;
}

sub _reload {
  my $cb = pop if ref( $_[-1] ) eq 'CODE';

  $INSTANCE->{config} = App::Environ::Config->instance->{'cow'};

  print __PACKAGE__ . " reloaded\n";

  if ( defined $cb ) {
    AE::postpone { $cb->() };
  }

  return;
}


sub _finalize {
  my $cb = pop if ref( $_[-1] ) eq 'CODE';

  undef $INSTANCE;

  print __PACKAGE__ . " finalized\n";

  if ( defined $cb ) {
    AE::postpone { $cb->() };
  }

  return;
}

1;


examples/lib/Dog.pm
package Dog;

use strict;
use warnings;

use App::Environ;
use App::Environ::Config;
use AnyEvent;
use Carp qw( croak );

App::Environ::Config->register( qw( dog.yml ) );

App::Environ->register( __PACKAGE__,
  initialize   => sub { __PACKAGE__->_initialize(@_) },
  reload       => sub { __PACKAGE__->_reload(@_) },
  'finalize:r' => sub { __PACKAGE__->_finalize(@_) },
);

my $INSTANCE;


sub instance {
  unless ( defined $INSTANCE ) {
    croak __PACKAGE__ . ' must be initialized first';
  }

  return $INSTANCE;
}

sub _initialize {
  my $class = shift;
  my $cb    = pop if ref( $_[-1] ) eq 'CODE';

  my $dog_config = App::Environ::Config->instance->{'dog'};

  $INSTANCE = {
    config    => $dog_config,
    init_args => [@_],
  };

  print __PACKAGE__ . " initialized\n";

  if ( defined $cb ) {
    AE::postpone { $cb->() };
  }

  return;
}

sub _reload {
  my $cb = pop if ref( $_[-1] ) eq 'CODE';

  $INSTANCE->{config} = App::Environ::Config->instance->{'dog'};

  print __PACKAGE__ . " reloaded\n";

  if ( defined $cb ) {
    AE::postpone { $cb->() };
  }

  return;
}


sub _finalize {
  my $cb = pop if ref( $_[-1] ) eq 'CODE';

  undef $INSTANCE;

  print __PACKAGE__ . " finalized\n";

  if ( defined $cb ) {
    AE::postpone { $cb->() };
  }

  return;
}

1;


lib/App/Environ/Config.pm
package App::Environ::Config;

use 5.008000;
use strict;
use warnings;

our $VERSION = '0.20';

use App::Environ;
use Config::Processor;
use Carp qw( croak );

my @REGD_CONFIG_SECTIONS;
my %CONFIG_SECTIONS_IDX;
my $CONFIG;
my $NEED_CONFIG_INIT = 1;

App::Environ->register( __PACKAGE__,
  initialize   => sub { __PACKAGE__->_initialize(@_) },
  reload       => sub { __PACKAGE__->_reload(@_) },
  'finalize:r' => sub { __PACKAGE__->_finalize(@_) },
);


sub register {
  my $class = shift;
  my @config_sections = @_;

  foreach my $config_section (@config_sections) {
    next if exists $CONFIG_SECTIONS_IDX{$config_section};

    $CONFIG_SECTIONS_IDX{$config_section} = 1;
    push( @REGD_CONFIG_SECTIONS, $config_section );

    $NEED_CONFIG_INIT = 1;
  }

  return;
}

sub instance {
  unless ( defined $CONFIG ) {
    croak __PACKAGE__ . ' must be initialized first';
  }

  return $CONFIG;
}

sub _initialize {
  my $class = shift;

  return unless $NEED_CONFIG_INIT;

  $class->_load;
  undef $NEED_CONFIG_INIT;

  return;
}

sub _reload {
  my $class = shift;

  $class->_load;

  return;
}

sub _finalize {
  undef $CONFIG;
  $NEED_CONFIG_INIT = 1;

  return;
}

sub _load {
  my @config_dirs;
  if ( defined $ENV{APPCONF_DIRS} ) {
    @config_dirs = split /:/, $ENV{APPCONF_DIRS};
  }
  my $interpolate_vars   = $ENV{APPCONF_INTERPOLATE_VARIABLES};
  my $process_directives = $ENV{APPCONF_PROCESS_DIRECTIVES};
  my $export_env         = $ENV{APPCONF_EXPORT_ENV};

  my $config_processor = Config::Processor->new(
    dirs => \@config_dirs,
    $interpolate_vars ? ( interpolate_variables => $interpolate_vars ) : (),
    $process_directives ? ( process_directives => $process_directives ) : (),
    $export_env         ? ( export_env         => $export_env )         : (),
  );

  $CONFIG = $config_processor->load(@REGD_CONFIG_SECTIONS);

  return;
}

1;
__END__
=head1 NAME

App::Environ::Config - Configuration files processor for App::Environ

=head1 SYNOPSIS

  use App::Environ;
  use App::Environ::Config;

  App::Environ::Config->register( qw( foo.yml bar.json ) );

  App::Environ->send_event('initialize');

  my $config = App::Environ::Config->instance;

=head1 DESCRIPTION

App::Environ::Config is the configuration files processor for App::Environ.
Allows get access to configuration tree from different application components.

The module registers in App::Environ three handlers for following events:
C<initialize>, C<reload> and C<finalize:r>.

=head1 METHODS

=head2 register( @config_sections )

The method registers configuration sections.

=head2 instance()

Gets reference to configuration tree.

=head1 ENVIRONMENT VARIABLES

You can control configuration file processing using environment variables.

=head2 APPCONF_DIRS

List of directories separated by ":" (colon), in which configuration processor
will search files. If the variable not specified, current directory will be
used.

=head2 APPCONF_INTERPOLATE_VARIABLES

Enables or disables variable interpolation in configurations files.
Enabled by default.

=head2 APPCONF_PROCESS_DIRECTIVES

Enables or disables directive processing in configuration files.
Enabled by default.

=head2 APPCONF_EXPORT_ENV

Enables or disables environment variables exporting to configuration tree.
If enabled, environment variables can be accessed by the key C<ENV> from the
configuration tree and can be interpolated into other configuration parameters.

Disabled by default.

=head1 SEE ALSO

L<App::Environ>, L<Config::Processor>

=head1 AUTHOR

Eugene Ponizovsky, E<lt>ponizovsky@gmail.comE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2016-2017, Eugene Ponizovsky, E<lt>ponizovsky@gmail.comE<gt>.
All rights reserved.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself.

=cut
